### PR TITLE
Video pr

### PR DESCRIPTION
**Pull request by Vagn99**
## Description

We have created routes for interacting with the GCP bucket(s) (including streaming video). We have also route for fetching lecture by id.

## Changes

New bucket routes (/bucket):

`GET /`  Returns list of everything in the bucket.
`GET /fileName`  Returns file in base64.
`GET /stream/:fileName`  Returns stream of file.

`DELETE /:fileName`  Deletes item in bucket.

`POST /upload`  Takes a `fileName` and `file` as formData. Uploads to GCP bucket.

## Related Issues

NIL

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [x] Performance impact of the changes has been evaluated, if relevant.

## Screenshots (if applicable)

image

image

image

## Notes for Reviewers
NIL

### CODE REVIEW

**Review by Mafusn — APPROVED** [review]
Looks good

**Comment by JackSMunn** [comment]
Please update changes to reflect the changes you have made.

File: routes/bucketRoutesGCP.js
Comment (Havkost) [inline routes/bucketRoutesGCP.js:1]: This file has serious security issues! 😱👻🎃
Right now it is possible to write arbitrary files to and even overwrite existing files on the server running the backend.

---

**POC attack against the backend running on localhost:8888**
Upload a file with a fileName equal to the path of the file you want to overwrite, in this case index.js just outside the _temp_bucketFiles directory (test.png is an image, but you can overwrite with any data):
`curl -X POST -H "Content-Type: multipart/form-data" -F "file=@test.png" -F "fileName=../index.js" http://localhost:8888/api/bucket/upload`

This has created the file in the bucket.\
Now we just need to read the file, which will write it to the local server, in turn overwriting the existing index.js:
`curl localhost:8888/api/bucket/download?fileName=../index.js`

**Review by Havkost — CHANGES_REQUESTED** [review]
## Don't merge - serious security issues
Please see my comment 👍

File: routes/bucketRoutesGCP.js
Comment (Vagn99) [reply routes/bucketRoutesGCP.js:1]: Should be fixed now! Please have a look🎉☀️

Comment (Havkost) [reply routes/bucketRoutesGCP.js:1]: Are you able to upload files to the server? I'm getting an error that uniqueFileName is not defined. If I define uniqueFileName in the form data, it says fileName is not defined, and if I supply both it goes back to saying that uniqueFileName is not defined.
Let's have a look at this together tomorrow - I'll be at uni at around 12:30 👍

Comment (Vagn99) [reply routes/bucketRoutesGCP.js:1]: I found the error, try it now 😎

**Comment by sofiiagran** [comment]
Looks good, but the dev branch has been updated since your PR was created. 
Please check and solve the new conflicts:))

**Review by sofiiagran — CHANGES_REQUESTED** [review]
Looks good, but the dev branch has been updated since your PR was created. 
Please check and solve the new conflicts:))

**Pull request merged**